builtin/logical/database/backend.go

```diff
@@ -166,46 +126,126 @@ func (b *databaseBackend) Role(ctx context.Context, s logical.Storage, roleName
+		}
+		if upgradeCh.Statements.RevocationStatements != "" {
+			stmts.RevocationStatements = []string{upgradeCh.Statements.RevocationStatements}
+		}
+		if upgradeCh.Statements.RollbackStatements != "" {
+			stmts.RollbackStatements = []string{upgradeCh.Statements.RollbackStatements}
+		}
+		if upgradeCh.Statements.RenewStatements != "" {
+			stmts.RenewStatements = []string{upgradeCh.Statements.RenewStatements}
+		}
+		result.Statements = stmts
+	case upgradeCh.OldStatements != nil:
+		var stmts dbplugin.Statements
+		if upgradeCh.OldStatements.CreationStatements != "" {
+			stmts.CreationStatements = []string{upgradeCh.OldStatements.CreationStatements}
+		}
+		if upgradeCh.OldStatements.RevocationStatements != "" {
+			stmts.RevocationStatements = []string{upgradeCh.OldStatements.RevocationStatements}
+		}
+		if upgradeCh.OldStatements.RollbackStatements != "" {
+			stmts.RollbackStatements = []string{upgradeCh.OldStatements.RollbackStatements}
+		}
+		if upgradeCh.OldStatements.RenewStatements != "" {
+			stmts.RenewStatements = []string{upgradeCh.OldStatements.RenewStatements}
+		}
+		result.Statements = stmts
 	}
 
 	return &result, nil
 }
 
 func (b *databaseBackend) invalidate(ctx context.Context, key string) {
-	b.Lock()
-	defer b.Unlock()
-
 	switch {
 	case strings.HasPrefix(key, databaseConfigPath):
 		name := strings.TrimPrefix(key, databaseConfigPath)
-		b.clearConnection(name)
+		b.ClearConnection(name)
 	}
 }
 
-// clearConnection closes the database connection and
-// removes it from the b.connections map.
-func (b *databaseBackend) clearConnection(name string) {
+func (b *databaseBackend) GetConnection(ctx context.Context, s logical.Storage, name string) (dbplugin.Database, error) {
+	b.RLock()
+	unlockFunc := b.RUnlock
+	defer func() { unlockFunc() }()
+
 	db, ok := b.connections[name]
 	if ok {
+		return db, nil
+	}
+
+	// Upgrade lock
+	b.RUnlock()
+	b.Lock()
+	unlockFunc = b.Unlock
+
+	db, ok = b.connections[name]
+	if ok {
+		return db, nil
+	}
+
+	config, err := b.DatabaseConfig(ctx, s, name)
+	if err != nil {
+		return nil, err
+	}
+
+	db, err = dbplugin.PluginFactory(ctx, config.PluginName, b.System(), b.logger)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Initialize(ctx, config.ConnectionDetails, true)
+	if err != nil {
 		db.Close()
+		return nil, err
+	}
+	b.connections[name] = db
+
+	return db, nil
+}
+
+// ClearConnection closes the database connection and
+// removes it from the b.connections map.
+func (b *databaseBackend) ClearConnection(name string) error {
+	b.Lock()
+	defer b.Unlock()
+	return b.clearConnection(name)
+}
+
+func (b *databaseBackend) clearConnection(name string) error {
+	db, ok := b.connections[name]
+	if ok {
+		if err := db.Close(); err != nil {
+			return err
+		}
 		delete(b.connections, name)
 	}
+	return nil
 }
 
-func (b *databaseBackend) closeIfShutdown(name string, err error) {
+func (b *databaseBackend) CloseIfShutdown(name string, err error) {
```

Comment:
How come these became exported?

Reply:
I use public/private more or less show show intent.  Maybe that is right, maybe not but I consider these public on the backend.  The backend itself is private, so it doesn't really matter.